feat(volume): CSI drivers

dancmeyers opened the pull request:
K8s volumes can be created from arbitrary CSI drivers, and can then be mounted on containers in the normal way. This PR adds support for creating CDK8s+ volumes from a CSI driver definition. This includes a generic `attributes` block to match the K8s resource, where driver-specific properties can be set.

The tests in this PR show an example for using the generic 'secrets store' CSI driver, which in turn defines its own CRD for interfacing with a specific provider (e.g. the one to fetch secrets from AWS Secrets Manager or SSM Parameter Store). I would expect this driver/provider to be the most common use-case in an AWS/EKS environment.

dancmeyers committed:
feat: Add volumes from Secrets Store

K8s has a CSI driver for generic 'secrets store' data. The CRD for that
defines a `SecretProviderClass` resource which has a parameters block to
pass arbitrary data to specific providers. Once a `SecretProviderClass`
object has been created, it can be mounted on a container as a volume.

Of specific interest to CDK8s is the AWS provider which allows secrets
to be automatically loaded from AWS Secrets Manager or SSM Parameter
store, including pulling specific keys out from JSON data, but all of
that logic is encapsulated in the `SecretProviderClass`. As such, this
modification should work with _any_ provider, not just the AWS one.

Signed-off-by: Dan Meyers <d.meyers@infinitefunspace.co.uk>

dancmeyers commented on src/volume.ts:
```diff
@@ -138,6 +139,33 @@ export class Volume extends Construct implements IStorage {
+   * @param options Options
+   */
+  public static fromSecretProvider(
+    scope: Construct, id: string, secretProvider: ss.SecretProviderClass, options: SecretProviderVolumeOptions = { }): Volume {
```
This will only type match if people use the `SecretProviderClass` from within CDK8s+, which is definitely not ideal. I'm thinking I'll do like the PVC ones and just pass in the name.

dancmeyers committed:
Refactor to not need concrete object

Requiring lib users ot pass in concrete objects that they may have
defined using a version of the class outside of CDK8s+ is a recipe for
pain. Given that we only actually need the `name`, I've refactored to
just expect that to be passed in instead.

Signed-off-by: Dan Meyers <d.meyers@infinitefunspace.co.uk>

iliapolo commented on src/volume.ts:
```diff
@@ -221,6 +221,32 @@ export class Volume extends Construct implements IStorage {
+   * @param className The name of the secret provider class to use to populate the volume.
+   * @param options Options
+   */
+  public static fromSecretProviderClass(scope: Construct, id: string, className: string, options: SecretProviderVolumeOptions = { }): Volume {
```
So my concern with this is that we are creating a coupling with specifications we have no control over, that are decoupled from the kubernetes spec itself. 

Imagine that `secrets-store` change either the schema (i.e the `secretProviderClass` property name or structure), or the driver name (i.e `secrets-store.csi.k8s.io`) - what would this function do? it would have to be incompatible with a certain version of the driver. 

I agree that this approach is better that accepting a `SecretProviderClass` object, but the underlying concern still holds true.

This is why integrations with third-party components should be implemented as third-party libraries (ideally published by the vendor itself). The problem of course is that such integrations are currently impossible because the `Volume` class is locked for extensions, but we will address this issue separately. 

In the meantime, I suggest we change this function to support the generic `csi` volume instead. Basically:

```suggestion
  public static fromCsi(scope: Construct, id: string, driver: string, options: CsiVolumeOptions = { }): Volume {
```

Where `CsiVolumeOptions` will contain an `attributes` property to hold the driver specific options.

dancmeyers replied:
Done!

dancmeyers committed:
Update to be generic for any CSI driver

The function is now called `fromCsi`, and accepts propeties such that it
should generate valid K8s resources for any CSI driver, although we
still reference integration with AWS SEcrets Manager / SSM as an
example.

Signed-off-by: Dan Meyers <d.meyers@infinitefunspace.co.uk>

iliapolo reviewed (approved):
Sweet, thanks!

dancmeyers commented:
@mergifyio refresh
